### ISSUE TITLE
Submitting ajax multi page form is submitting valid submissions as incomplete and not using the redirection settings

### ISSUE THREAD

**Issue by jamesmacwhite** [verbb/formie#2160]
### Describe the bug

I have a two-page form which is exhibiting a strange behaviour with valid submissions which end up being saved as incomplete. What's even stranger is in some cases the form submits fine without issues for me and redirects to the set confirmation page entry, but others have the issue described below and I can't replicate consistently.

The behaviour we are seeing is valid submissions after submitting are causing a scroll to the top of the form as if there is a validation error, however there is no error on any field, I found the payload being sent was a success response, but our redirection settings have not been used. This is the payload being sent:

```JSON
{
    "success": true,
    "submissionId": 525876,
    "nextPageId": 44,
    "nextPageIndex": 1,
    "isFinalPage": false,
    "totalPages": 2,
    "redirectUrl": "https://staging.nottinghamcollege.ac.uk/student-life/support-funding/exam-services/private-exam-entry-service/exam-service-enquiry-confirmation",
    "submitActionMessage": "<p>Submission saved.</p>",
    "events": []
}
```

The redirection settings set on the form:

![image](https://github.com/user-attachments/assets/68e03fcd-a76f-4d3f-a8ff-52395eaf54f2)

Looking at the payload the `isFinalPage` seems suspect because this is a two page form and is the last page, but the JSON data is suggesting it is not.

We use the Ajax client method, because of being a multi-page, when set to server any errors or issues jump the scroll position away from the form on reload, which isn't ideal.

The data appears to be saved but as incomplete and doesn't trigger the email notification configured, because of the incomplete status.

Not sure if there's a bug here or not. Possibly related to Craft Cloud maybe? I don't know.

Any thoughts?

### Steps to reproduce

1. Setup multi-page form with two pages
2. Set a confirmation page as a redirect URL
3. Setup page one with a continue button
4. Setup page two with some data capture fields
5. Submit the form from the final page.
6. Scroll to the top behaviour occurs with JSON payload stating its not the last page.

### Form settings

- Multi-page form: Yes
- Submission Method: Ajax
- Client-side Validation: Yes
- Custom Form Templates: No

### Craft CMS version

5.5.3

### Plugin version

3.0.12

### Multi-site?

No

**Comment by rickmerkelbach** [comment]
I’m seeing this exact issue on multiple projects as well, the main thing they seem to have in common is they’re all using Craft 5.5.0.1 or above. Almost all submissions “end” at the second to last page.

As a short term fix we’re converting them to single page forms.

**Comment by engram-design** [comment]
A few people seem to be reporting this, despite not being able to replicate it on my end. These aren't on Craft Cloud.

It seems like the session-handling that remembers which page you're on is getting messed up. While I can't uncover what's going on at the moment, I can provide a solution that take sessions out of the equation, and the page submission is taken control of at the client-end. 

It's something I've been meaning to do at some point anyway. It may very well even become the default handler. There's all sort of issues with Blitz and other caching mechanisms relying on sessions, that just won't work.

The only real caveat to using non-session handling, is that if you refresh your page in the middle of a form submission, it will reset the current page. That's because everything is in-memory for the current request.

To get this early, run `composer require verbb/formie:"dev-craft-5 as 3.0.12"`.

Your only change should be adding in your templates:

```twig
{% do form.setStorageBehaviour('memory') %}
```

To opt-into this behaviour. Right now, we only add storage support for the current page, but will expand to editing submissions and snapshot data (for form and fields).

**Comment by jamesmacwhite** [comment]
Thanks Josh. I will admit I struggled to replicate this reliably as well, but managed to see it first hand on other people's devices, then managed to make it happen once on my own machine, but can't do reliably sadly. I've narrowed it down to multi page forms at least, as we aren't seeing issues with single page forms.

We are including forms using a matrix block with the `{{ craft.formie.renderForm(form) }}` Twig variable. Is adding the `{% do form.setStorageBehaviour('memory') %}` line OK in this block to opt-in?

I guess the alternative is like @rickmerkelbach said is convert it to a single page as well temporarily, as it is specifically multi-page specific.

**Comment by engram-design** [comment]
Yep, go ahead and add that just above that call - keen to see if that helps, but I dare say it should.

Providing a different storage behaviour is a long-term goal (just started it early to address this), so it's not going anywhere, as opposed to a temporary fix. As for _why_ sessions have all of a sudden caused us an issue it's tough to say. I know Craft Cloud's sessions are a bit wild, so I wouldn't be surprised if it's not exactly compatible.

**Comment by jamesmacwhite** [comment]
We'll give it a go and report back. Thank you. We need that Craft Cloud path fix anyway, so I'll update our composer to dev-craft-5 to check things out.

@rickmerkelbach Just for additional info are any of the projects you mentioned hosted on Craft Cloud by any chance just for comparison?

**Comment by rickmerkelbach** [comment]
@engram-design Thanks! We’re going to try this as well.

@jamesmacwhite No Craft Cloud here, just Craft 5.5.0.1 - 5.5.3 installs on our own servers (no Redis session handling).

**Comment by jamesmacwhite** [comment]
@rickmerkelbach Thanks for the info!

**Comment by engram-design** [comment]
@rickmerkelbach Maybe something about session handling has changed in 5.5+, just makes it hard to troubleshoot being unable to replicate it on my end, sorry!

**Comment by jamesmacwhite** [comment]
Just to report back that deploying the storage behaviour hasn't seemed to help us. We are going to remove multi-page forms for now and go back to one page, it seems too unreliable at the moment.

Totally understand it is hard with reproducing it, I had difficulty myself!

**Comment by jamesmacwhite** [comment]
If you'd like any help trying to reproduce, willing to help. I know of at least one scenario, but getting it to consistently happen was my problem, but seeing the payload info, gave me an idea of things going wrong.

I don't know if server side vs Ajax has relevance as well. Interesting though, that @rickmerkelbach mentioned it was happening on environments without load balancing/session customisation i.e. Redis.

Could flag with the Craft CMS dev team? I scanned the changelog for 5.5.0 could not see anything that stands out re: sessions.

**Comment by engram-design** [comment]
So that's incredibly strange that the in-memory option isn't working, we shouldn't even be using sessions in that instance. I can spin up a Craft Cloud project to test further. Anything about your setup out of the norm?

**Comment by jamesmacwhite** [comment]
Nothing different that I'd determine to be a major area here. Although we were working quite fast to fix the issue with a live form, so I will perhaps go back through and test again to double check with a bit less pressure on. What didn't help at the same time, we switched out the captcha solution and Craft Cloud did something weird where by it dropped two environment variables deployed which happens to be the reCAPTCHA site and secret keys. That wouldn't have helped with submit tests I'd imagine!

I'd be interested to know from @rickmerkelbach if you had success applying the switch to in-memory, given your setup sounded pretty vanilla, so less variables to potentially deal with! If so, I'll revisit our scenario and test again.

**Comment by jedso** [comment]
Hey @engram-design,

So can confirm also having this issue with an AJAX multi-page form with Formie v3.0.15 and Craft 5.5.6.1, and now have a reproducible example that consistently hits this bug (at least on my end). In our case, we have reCAPTCHA v3 enabled on the form, although it seems like this bug is happening regardless of if this is turned on in the form or not.

When logged in as admin, the bug never happens for us. When opening the form in a new incognito window (and therefore logged out), the final page consistently submits as incomplete. E.g. here's our form with reCAPTCHA turned off:

https://github.com/user-attachments/assets/a4681897-30d5-4144-a195-8cce54b5a3f4

And then a working example in the same incognito window, but logged in as admin:

https://github.com/user-attachments/assets/7aae9a9a-2d3a-48ae-ac54-ebdff2123c48

This behavior happens consistently for us in Firefox and Chrome, tested on both Debian and MacOS.

Now it's worth noting that in our case, `{% do form.setStorageBehaviour('memory') %}` _does_ fix this issue, but we're using a vanilla setup rather than Craft Cloud.

I'll send in a support request with the exact form config, but if you aren't able to reproduce it locally and I can probably give you an internet-facing URL that produces the bug.

**Comment by engram-design** [comment]
Thanks for putting that all together, I'll respond to your support request! Something fishy is certainly going on with sessions it seems...

**Comment by engram-design** [comment]
Related to https://github.com/verbb/formie/issues/2194#issuecomment-2566771363 I _think_ I've gotten to the bottom of this without the need to switch storage behaviours (but you can continue to use `memory` if you like, it's not going anywhere).

To get this early, run `composer require verbb/formie:"dev-craft-5 as 3.0.16"`.

**Comment by jamesmacwhite** [comment]
@engram-design Great news, thanks for the update. I'll do some testing and report back on behalf of any Craft Cloud hosted people.

**Comment by engram-design** [comment]
Updated in [3.0.17](https://github.com/verbb/formie/releases/tag/3.0.17)
